daemon/ipcache.go

```diff
@@ -0,0 +1,147 @@
+// Copyright 2016-2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"net"
+
+	"github.com/cilium/cilium/api/v1/models"
+	. "github.com/cilium/cilium/api/v1/server/restapi/policy"
+	"github.com/cilium/cilium/pkg/api"
+	"github.com/cilium/cilium/pkg/identity"
+	"github.com/cilium/cilium/pkg/ipcache"
+	monitorAPI "github.com/cilium/cilium/pkg/monitor/api"
+
+	"github.com/go-openapi/runtime/middleware"
+)
+
+// OnIPIdentityCacheChange listens to ipcache updates and forwards them to the monitor
+func (d *Daemon) OnIPIdentityCacheChange(modType ipcache.CacheModification, cidr net.IPNet,
```

Comment:
How tightly do we want to bind this to the actual change in datapath state?

Currently this is implemented as a separate notification that is triggered from the ipcache, which may occur before or after the actual change to BPF, with no timestamp reference as to whether the change in datapath behaviour just occurred or is about to occur. I *suspect* that it will happen just after the BPF is plumbed because the listeners are a slice and this likely implies ordering, but I don't know if this is guaranteed.

It could be more closely tied by having some integration into `pkg/datapath/ipcache`, but then I don't know in practical terms how much more accurate we will get and whether it's worth the code complexity. At some point, the syscall to modify the datapath state will occur, and the notification needs to be triggered before/after. But figured I would ask just to see if you thought about that.

Reply:
Good point, thanks for bringing this up! I did not consider that. The current design was mostly chosen to avoid much code coupling.

I guess without actually sending some form of ipcache update notification though the perf ring buffer there cannot be any strict order guarantees for cilium monitor clients between ipcache notifications and trace notifications (which is what we ideally want as a cilium monitor client). This in turns would mean we have to send the notification from a BPF program? That sounds like a lot of complexity.

For now, I will change the PR to emit the monitor notification in `pkg/datapath/ipcache`, as it seems relatively straight-forward to do.

Reply:
(As a side note, thanks as always for the amazing code review - much appreciated!)

Reply:
Agreed that you would need a single channel, but I don't know that there's a really good way to emit an event into the perf event channel when the element is added based on the syscall; maybe something could be done with tracepoints but for now this is more than necessary.